src/cogent3/core/new_alignment.py

```diff
@@ -0,0 +1,449 @@
+        else:
+            start, stop = self.start, self.stop
+
+        data["init_args"]["seq"] = self.seq[start:stop]
+        data["init_args"]["offset"] = int(self.parent_start)
+        return data
+
+    @classmethod
+    def from_rich_dict(cls, data: dict):
+        init_args = data.pop("init_args")
+        if "offset" in data:
+            init_args["offset"] = data.pop("offset")
+        return cls(**init_args)
+
+    def copy(self, sliced: bool = False):
+        """returns copy
+
+        Parameters
+        ----------
+        sliced
+            if True, the underlying sequence is truncated and the start/stop
+            adjusted
+        """
+        if not sliced:
+            return self.__class__(
+                seq=self.seq,
+                start=self.start,
+                stop=self.stop,
+                step=self.step,
+                offset=self.offset,
+                seqid=self.seqid,
+                seq_len=self.seq_len,
+            )
+        return self.from_rich_dict(self.to_rich_dict())
+
+
+@dataclass
+class SeqData:
+    __slots__ = ("_data", "_alphabet", "_name_order", "_seq_maker")
+
+    def __init__(
+        self,
+        data: dict[str, SeqTypes],
+        alphabet: CharAlphabet,
+        name_order: Optional[Union[tuple[str], None]] = None,
+        seq_maker: Optional[type] = None,
+    ):
+        self._alphabet = alphabet
+        self._name_order = process_name_order(data, name_order)
+        self._seq_maker = seq_maker
+        # Convert to moltype alphabet indicies
+        self._data: dict[str, numpy.ndarray] = {
+            k: seq_index(v, self._alphabet) for k, v in data.items()
+        }
+
+    @property
+    def set_seq_maker(self) -> type:
+        return self._seq_maker
+
+    @set_seq_maker.setter
+    def set_seq_maker(self, value: tuple[type, CharAlphabet]) -> type:
+        seq_maker, alphabet = value
+        if alphabet != self._alphabet:
+            raise ValueError("Alphabets must be the same")
+        assert alphabet == self._alphabet
+        self._seq_maker = seq_maker
+
+    def _make_seq(self, seqid) -> Union[SeqDataView, type]:
+        seqdataview = self.get_seq_view(seqid=seqid)
+        if self._seq_maker is None:
+            return seqdataview
+        return self.set_seq_maker(seqdataview, seqid=seqid)
+
+    @singledispatchmethod
+    def __getitem__(self, value: Union[str, int]) -> SeqDataView:
+        raise NotImplementedError(f"__getitem__ not implemented for {type(value)}")
+
+    @__getitem__.register
+    def _(self, value: str) -> SeqDataView:
+        return self._make_seq(seqid=value)
+
+    @__getitem__.register
+    def _(self, value: int) -> SeqDataView:
+        seqid = self._name_order[value]
+        return self._make_seq(seqid=seqid)
+
+    def get_seq_array(
+        self, *, seqid: str, start: int = None, stop: int = None
+    ) -> numpy.ndarray:
+        return self._data[seqid][start:stop]
```

Comment:
Since `get_seq_array` is a class method, would it be better not to require all the args to be provided (like `stop` and `start`) and rely on `self.stop`, etc, or do we foresee wanting flexible indexing?